Eliminate `ObligationCauseData`

nnethercote opened the pull request:
This makes `Obligation` two words bigger, but avoids allocating a lot of the time.
    
I previously tried this in #73983 and it didn't help much, but local timings look more promising now.

r? @ghost

nnethercote commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit ddb96eb0b1587c670438b0e577eb709c90e29447 with merge 220db1ba07e4e16cdbe3889232735221018fe852...
<!-- homu: {"type":"TryBuildStarted","head_sha":"ddb96eb0b1587c670438b0e577eb709c90e29447","merge_sha":"220db1ba07e4e16cdbe3889232735221018fe852"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4500573914?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/ddb96eb0b1587c670438b0e577eb709c90e29447/checks/4500573914/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
failures:

---- [mir-opt] mir-opt/issue-41888.rs stdout ----
22 
23     bb0: {
24         _9 = const false;                // scope 0 at $DIR/issue-41888.rs:7:9: 7:10
-         _8 = const false;                // scope 0 at $DIR/issue-41888.rs:7:9: 7:10
26         _7 = const false;                // scope 0 at $DIR/issue-41888.rs:7:9: 7:10
+         _8 = const false;                // scope 0 at $DIR/issue-41888.rs:7:9: 7:10
27         StorageLive(_1);                 // scope 0 at $DIR/issue-41888.rs:7:9: 7:10
28         StorageLive(_2);                 // scope 1 at $DIR/issue-41888.rs:8:8: 8:14
29         _2 = cond() -> [return: bb1, unwind: bb11]; // scope 1 at $DIR/issue-41888.rs:8:8: 8:14

thread '[mir-opt] mir-opt/issue-41888.rs' panicked at 'Actual MIR output differs from expected MIR output /checkout/src/test/mir-opt/issue_41888.main.ElaborateDrops.after.mir', src/tools/compiletest/src/runtest.rs:3360:25


failures:
    [mir-opt] mir-opt/issue-41888.rs
    [mir-opt] mir-opt/issue-41888.rs

test result: FAILED. 162 passed; 1 failed; 3 ignored; 0 measured; 0 filtered out; finished in 4.56s



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/mir-opt" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/mir-opt" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "mir-opt" "--mode" "mir-opt" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-12/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "12.0.0" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker engine executionengine extensions filecheck frontendopenacc frontendopenmp fuzzmutate globalisel hellonew hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interfacestub interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcjit orcshared orctargetprocess passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--channel" "nightly" "--color" "always"


Build completed unsuccessfully in 0:13:58
```

</details>

nnethercote commented:
@bors try

bors commented:
:hourglass: Trying commit 748ccba835c1336c51462525d136e2314289a284 with merge 498fbcc6241318ce8d97f067af15d4fba425d18b...
<!-- homu: {"type":"TryBuildStarted","head_sha":"748ccba835c1336c51462525d136e2314289a284","merge_sha":"498fbcc6241318ce8d97f067af15d4fba425d18b"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/4503349360?check_suite_focus=true)
Build commit: 498fbcc6241318ce8d97f067af15d4fba425d18b (`498fbcc6241318ce8d97f067af15d4fba425d18b`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/4503349360?check_suite_focus=true"},"merge_sha":"498fbcc6241318ce8d97f067af15d4fba425d18b"} -->

rust-timer commented:
Queued 498fbcc6241318ce8d97f067af15d4fba425d18b with parent f7fd79ac1d485ab47b62146f6dafed4aad5d9c6d, future [comparison URL](https://perf.rust-lang.org/compare.html?start=f7fd79ac1d485ab47b62146f6dafed4aad5d9c6d&end=498fbcc6241318ce8d97f067af15d4fba425d18b).

jyn514 commented on compiler/rustc_middle/src/traits/mod.rs:
```diff
@@ -174,6 +162,11 @@ impl<'tcx> ObligationCause<'tcx> {
+
+    #[inline]
+    pub fn code(&self) -> &ObligationCauseCode<'tcx> {
+        self.code.as_deref().unwrap_or(&MISC_OBLIGATION_CAUSE_CODE)
```
Hmm, are you sure the compiler is smart enough to avoid dereferencing the pointer when the code is `None`?

nnethercote replied:
Sorry, not sure I follow. Is this a comment about correctness, or something else?

The old code had a very similar construct.

jyn514 replied:
No, just a comment about performance - it should be possible to avoid dereferencing `code` when it's `None`, but I'm not sure whether the compiler does that in practice.

rust-timer commented:
Finished benchmarking commit (498fbcc6241318ce8d97f067af15d4fba425d18b): [comparison url](https://perf.rust-lang.org/compare.html?start=f7fd79ac1d485ab47b62146f6dafed4aad5d9c6d&end=498fbcc6241318ce8d97f067af15d4fba425d18b).

**Summary**: This change led to large relevant mixed results 🤷 in compiler performance.
- Large improvement in instruction counts (up to -3.8% on `incr-unchanged` builds of `deep-vector`)
- Small regression in instruction counts (up to 0.7% on `incr-full` builds of `deeply-nested`)

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR led to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf +perf-regression

nnethercote commented:
Lots more wins than losses among the instruction counts, plus it's a tiny reduction in number of lines of code. I think this is worth going ahead with.

r? @Mark-Simulacrum

Mark-Simulacrum commented on compiler/rustc_middle/src/traits/mod.rs:
```diff
@@ -122,46 +96,60 @@ pub struct ObligationCauseData<'tcx> {
     fn hash<H: Hasher>(&self, state: &mut H) {
         self.body_id.hash(state);
         self.span.hash(state);
         std::mem::discriminant(&self.code).hash(state);
```
This is now hashing an Option -- probably you actually want `self.code()` here? Alternatively, my guess is that hashing the option discriminant is approximately equivalent to just not hashing the code at all, so maybe we should just drop the hash of the discriminant entirely?

nnethercote replied:
Good catch. It turns out that 99.99%+ of the time the `self.code()` is a `ImplDerivedObligation`. Not sure why, but no point hashing either discriminant. I will remove it.

Mark-Simulacrum commented:
I agree that these wins exceed the regressions, so this is worth it.

I am wondering, though, if it makes sense to try to go further here. In particular, it seems to me that many ObligationCauseCode's are or could be <= one pointer in "size", in which case `Lrc<...>` is adding an allocation that's not really necessary.

For example, just by moving the discriminant into the Obligation (one extra byte), we'd be able to represent a lot of the variants inline -- trivially all of the ones that have no data attached, and in theory those which have data that's only usize wide (e.g., a single DefId). I don't off-hand know what the distribution there looks like -- this PR adds a comment suggesting this covers 60% of cases with just MiscObligation, but it's worth noting that we keep adding new obligation cause codes, so presumably that percentage is only going to decrease over time. I'm not sure what the distribution on the rest of the codes is like.

Further optimization seems fine to leave for future PRs, but I would like the comment on the Option discriminant hashing addressed; r=me with that done (another perf run may be in order).

nnethercote commented:
I have updated with the suggested change.

> I am wondering, though, if it makes sense to try to go further here.

Interesting idea. `Obligation` contains an `ObligationCause`, which contains an `ObligationCauseCode`. `Obligation` has a `recursive_depth` field that is a `usize` that could easily shrink to a `u32` or even `u16`. That's the only place I can see where there's room to maneuver. So we could effectively split `ObligationCauseCode` in half: one part for all the simple variants, one part for all the complex variants. For simple variants we could squeeze it next to `recursion_limit` and `code` would be `None`. For complex variants `code` would be `Some`. (I've just restated your suggestion with a little more specifics.)

The downside is that `ObligationCauseCode` is split in two. I guess we'd have three types: one for simple variants, one for complex variants, and the full type that gets assembled from the former two on demand. But as you say, it would require some measurement of how common the other "simple" variants are.

nnethercote commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit 47d6bdfd7b00e28839d261f1c7a9de190560079f with merge ee0165f75bf861e6b1d07c3312a2e8491262b13d...
<!-- homu: {"type":"TryBuildStarted","head_sha":"47d6bdfd7b00e28839d261f1c7a9de190560079f","merge_sha":"ee0165f75bf861e6b1d07c3312a2e8491262b13d"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/4530618992?check_suite_focus=true)
Build commit: ee0165f75bf861e6b1d07c3312a2e8491262b13d (`ee0165f75bf861e6b1d07c3312a2e8491262b13d`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/4530618992?check_suite_focus=true"},"merge_sha":"ee0165f75bf861e6b1d07c3312a2e8491262b13d"} -->

rust-timer commented:
Queued ee0165f75bf861e6b1d07c3312a2e8491262b13d with parent d594910a2da12f158477b4c7281716f535cfa3de, future [comparison URL](https://perf.rust-lang.org/compare.html?start=d594910a2da12f158477b4c7281716f535cfa3de&end=ee0165f75bf861e6b1d07c3312a2e8491262b13d).

rust-timer commented:
Finished benchmarking commit (ee0165f75bf861e6b1d07c3312a2e8491262b13d): [comparison url](https://perf.rust-lang.org/compare.html?start=d594910a2da12f158477b4c7281716f535cfa3de&end=ee0165f75bf861e6b1d07c3312a2e8491262b13d).

**Summary**: This change led to moderate relevant improvements 🎉 in compiler performance.
- Moderate improvement in instruction counts (up to -1.1% on `full` builds of `deep-vector`)

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR led to changes in compiler perf.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf -perf-regression

nnethercote commented:
Similar results on this new perf run, apart from `deep-vector debug incr-unchanged`, but I suspect that was just noise in the first run.

Mark-Simulacrum commented:
@bors r+ 

Yeah -- going to approve this collection of wins for now, can always continue iterating.

bors commented:
:pushpin: Commit 47d6bdfd7b00e28839d261f1c7a9de190560079f has been approved by `Mark-Simulacrum`

<!-- @bors r=Mark-Simulacrum 47d6bdfd7b00e28839d261f1c7a9de190560079f -->
<!-- homu: {"type":"Approved","sha":"47d6bdfd7b00e28839d261f1c7a9de190560079f","approver":"Mark-Simulacrum"} -->

bors commented:
:hourglass: Testing commit 47d6bdfd7b00e28839d261f1c7a9de190560079f with merge d985d42bb4d0910b017e761acb7125427ce7aef2...
<!-- homu: {"type":"BuildStarted","head_sha":"47d6bdfd7b00e28839d261f1c7a9de190560079f","merge_sha":"d985d42bb4d0910b017e761acb7125427ce7aef2"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/4568376460?check_suite_focus=true)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/4568376460?check_suite_focus=true","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/4568358843?check_suite_focus=true) [(plain)](https://github.com/rust-lang-ci/rust/commit/d985d42bb4d0910b017e761acb7125427ce7aef2/checks/4568358843/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_save_analysis v0.0.0 (/checkout/compiler/rustc_save_analysis)
    Checking rustc_codegen_ssa v0.0.0 (/checkout/compiler/rustc_codegen_ssa)
    Checking rustc_resolve v0.0.0 (/checkout/compiler/rustc_resolve)
    Checking rustc_trait_selection v0.0.0 (/checkout/compiler/rustc_trait_selection)
error[E0616]: field `code` of struct `rustc_middle::traits::ObligationCause` is private
   --> compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:889:78
    |
889 |         if let ObligationCauseCode::AwaitableExpr(hir_id) = obligation.cause.code.peel_derives() {
    |
    |
help: a method `code` also exists, call it with parentheses
    |
889 |         if let ObligationCauseCode::AwaitableExpr(hir_id) = obligation.cause.code().peel_derives() {


error[E0599]: no method named `peel_derives` found for enum `Option` in the current scope
   --> compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:889:83
    |
889 |         if let ObligationCauseCode::AwaitableExpr(hir_id) = obligation.cause.code.peel_derives() {
    |                                                                                   ^^^^^^^^^^^^ method not found in `Option<Arc<rustc_middle::traits::ObligationCauseCode<'tcx>>>`
    Checking rustc_codegen_llvm v0.0.0 (/checkout/compiler/rustc_codegen_llvm)
Some errors have detailed explanations: E0599, E0616.
For more information about an error, try `rustc --explain E0599`.
error: could not compile `rustc_trait_selection` due to 2 previous errors
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #92099) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

nnethercote commented:
I rebased to fix the conflict. While there, I added a `clean_code()` function that trivially avoids some additional allocations relating to the `parent_code` field, that helps with #87012.

@bors r=Mark-simulacrum

bors commented:
:pushpin: Commit d7ca962282ccc09fad6c8193fa3a474e339097c5 has been approved by `Mark-simulacrum`

<!-- @bors r=Mark-simulacrum d7ca962282ccc09fad6c8193fa3a474e339097c5 -->
<!-- homu: {"type":"Approved","sha":"d7ca962282ccc09fad6c8193fa3a474e339097c5","approver":"Mark-simulacrum"} -->

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4576424980?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/d7ca962282ccc09fad6c8193fa3a474e339097c5/checks/4576424980/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_save_analysis v0.0.0 (/checkout/compiler/rustc_save_analysis)
    Checking rustc_codegen_ssa v0.0.0 (/checkout/compiler/rustc_codegen_ssa)
    Checking rustc_resolve v0.0.0 (/checkout/compiler/rustc_resolve)
    Checking rustc_trait_selection v0.0.0 (/checkout/compiler/rustc_trait_selection)
error[E0616]: field `code` of struct `rustc_middle::traits::ObligationCause` is private
   --> compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:889:78
    |
889 |         if let ObligationCauseCode::AwaitableExpr(hir_id) = obligation.cause.code.peel_derives() {
    |
    |
help: a method `code` also exists, call it with parentheses
    |
889 |         if let ObligationCauseCode::AwaitableExpr(hir_id) = obligation.cause.code().peel_derives() {


error[E0599]: no method named `peel_derives` found for enum `Option` in the current scope
   --> compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:889:83
    |
889 |         if let ObligationCauseCode::AwaitableExpr(hir_id) = obligation.cause.code.peel_derives() {
    |                                                                                   ^^^^^^^^^^^^ method not found in `Option<Arc<rustc_middle::traits::ObligationCauseCode<'tcx>>>`
    Checking rustc_codegen_llvm v0.0.0 (/checkout/compiler/rustc_codegen_llvm)
Some errors have detailed explanations: E0599, E0616.
For more information about an error, try `rustc --explain E0599`.
error: could not compile `rustc_trait_selection` due to 2 previous errors
```

</details>

nnethercote commented:
@bors r=Mark-Simulacrum

bors commented:
:pushpin: Commit f09b1facd04250135d9ad0434c73378a6c08f415 has been approved by `Mark-Simulacrum`

<!-- @bors r=Mark-Simulacrum f09b1facd04250135d9ad0434c73378a6c08f415 -->
<!-- homu: {"type":"Approved","sha":"f09b1facd04250135d9ad0434c73378a6c08f415","approver":"Mark-Simulacrum"} -->

bors commented:
:hourglass: Testing commit f09b1facd04250135d9ad0434c73378a6c08f415 with merge ed7a2068432b62880bca89cd69546682a2e192cd...
<!-- homu: {"type":"BuildStarted","head_sha":"f09b1facd04250135d9ad0434c73378a6c08f415","merge_sha":"ed7a2068432b62880bca89cd69546682a2e192cd"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/4578077553?check_suite_focus=true)
Approved by: Mark-Simulacrum
Pushing ed7a2068432b62880bca89cd69546682a2e192cd to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Mark-Simulacrum","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/4578077553?check_suite_focus=true"},"merge_sha":"ed7a2068432b62880bca89cd69546682a2e192cd"} -->

nnethercote commented:
> So we could effectively split `ObligationCauseCode` in half: one part for all the simple variants, one part for all the complex variants.

I did some quick measurements and the potential for improvement here seems very low. The simple variants account for a small fraction (~5% or less) of the occurrences seen in practice (while compiling `std`).

rust-timer commented:
Finished benchmarking commit (ed7a2068432b62880bca89cd69546682a2e192cd): [comparison url](https://perf.rust-lang.org/compare.html?start=e95e084a14870a718c712936ab5a8f8cd0159485&end=ed7a2068432b62880bca89cd69546682a2e192cd).

**Summary**: This change led to small relevant mixed results 🤷 in compiler performance.
- Small improvement in instruction counts (up to -0.9% on `full` builds of `coercions`)
- Small regression in instruction counts (up to 0.9% on `full` builds of `wg-grammar`)

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).



**Next Steps**: If you can justify the regressions found in this perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please open an issue or create a new PR that fixes the regressions, add a comment linking to the newly created issue or PR, and then add the `perf-regression-triaged` label to this PR.

@rustbot label: +perf-regression
